### PR TITLE
Dedicated `sys` types

### PR DESCRIPTION
**Pull request by marcolink**
## Summary
Dedicated types for different entities.

## Description
Given the different shapes of the `sys` field between the different entities, we want to reflect that also on a type level.

### CODE REVIEW

**Review by ruderngespra — APPROVED** [review]
LGTM! Just one follow up question: The `baseSys` here has a `locale` property, but tags don't have this. Wouldn't this become a bit annoying in the future?

**Comment by marcolink** [comment]
> LGTM! Just one follow up question: The `baseSys` here has a `locale` property, but tags don't have this. Wouldn't this become a bit annoying in the future?
> 
> But probably we should not worry about this at this early stage with no tags available in the cda?

@ruderngespra You're right!
Please check the reshaped types with a very basic `BaseSys`

**Review by damienxy — APPROVED** [review]

File: lib/common-types.ts [hunk 1]
```diff
@@ -1,91 +1,126 @@
...
   revision?: number
-  space?: {
-    sys: SpaceLink
-  }
-  contentType: {
-    sys: ContentTypeLink
-  }
+  space?: { sys: SpaceLink }
+  environment: { sys: EnvironmentLink }
+  locale: string
```
Comment (TimBeyer) [inline lib/common-types.ts:90]: I think `locale` is only present when you don't request via `locale=*`.
So probably we will have to split this out as well.

File: lib/common-types.ts [hunk 2]
```diff
@@ -1,91 +1,126 @@
...
   updatedAt: string
-  locale: string
   revision?: number
-  space?: {
-    sys: SpaceLink
-  }
-  contentType: {
-    sys: ContentTypeLink
-  }
+  space?: { sys: SpaceLink }
```
Comment (TimBeyer) [inline lib/common-types.ts:88]: Do we know when `space` is actually not present?  
Those optional properties will always force people to check the existence in their code.
Maybe we already know if it exists or not?

File: lib/common-types.ts [hunk 3]
```diff
@@ -1,91 +1,126 @@
...
+interface BaseSys {
   type: string
   id: string
+}
+
+type EntitySys = {
   createdAt: string
   updatedAt: string
-  locale: string
   revision?: number
```
Comment (TimBeyer) [inline lib/common-types.ts:87]: When is `revision` not present?
I think for entries, assets and content types it's always there.

File: lib/common-types.ts [hunk 4]
```diff
@@ -1,91 +1,126 @@
...
+export interface Sys extends EntitySys {
+  contentType: { sys: ContentTypeLink }
 }
-export interface Link<T extends string> {
+
+export type AssetSys = EntitySys
+
+export type SpaceSys = { type: 'Space' } & BaseSys
+
+export type ContentTypeSys = {
```
Comment (TimBeyer) [inline lib/common-types.ts:101]: The pattern in the preceding types was to use `interface Foo extends Bar`.
Maybe we should do that here in the new sys types  as well for consistency.

**Pull request merged**